compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -360,13 +391,37 @@ impl<'a> Parser<'a> {
             }
             Err(err) => return Err(err),
         };
-        let ty = if self.eat(&token::Semi) {
-            TyKind::Array(elt_ty, self.parse_anon_const_expr()?)
-        } else {
-            TyKind::Slice(elt_ty)
-        };
-        self.expect(&token::CloseDelim(token::Bracket))?;
-        Ok(ty)
+        let size_const =
+            if self.eat(&token::Semi) { Some(self.parse_anon_const_expr()?) } else { None };
+
+        match self.expect(&token::CloseDelim(token::Bracket)) {
+            Ok(_) => Ok(match size_const {
+                Some(size_const) => TyKind::Array(elt_ty, size_const),
+                None => TyKind::Slice(elt_ty),
+            }),
+            Err(mut err1)
+                if recover_array_type == RecoverArrayType::Yes
+                    && self.look_ahead(0, |t| t.kind == token::Comma)
+                    && size_const.is_none() =>
+            {
+                // Possibly used ',' instead of ';' in slice type syntax, try to recover
+                self.bump();
+                let snapshot = self.clone();
+                match self.parse_anon_const_expr() {
+                    Ok(array_size) => {
+                        err1.emit();
+                        self.expect(&token::CloseDelim(token::Bracket))?;
```

Comment:
This line is another piece of signal. If the closing bracket isn't there, then you have no confidence that the recovery is actually correct and you should return `Err(err1)`. So instead of `expect` you can use `eat`.

Reply:
This code is now rewritten using a 4-way match.